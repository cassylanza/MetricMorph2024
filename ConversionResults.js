import React, { Component } from 'react'
import { TouchableOpacity, Text, View, StyleSheet, ScrollView } from 'react-native'
import { Actions } from 'react-native-router-flux'


// formats the calculated time into hours, minutes, and seconds
// returns default of "hh:mm:ss" if input is invalid (i.e. not a number
// or outside of the specified range)
export function formatResult(timeInSeconds) {

  let maxNumber = 362399; // formats to 99:59:59

  if (isNaN(timeInSeconds) || typeof timeInSeconds !== 'number') {
      return "hh:mm:ss";
  }

  if(timeInSeconds < 0 || timeInSeconds > maxNumber) {
    return "hh:mm:ss";
  }

  let hours = parseInt((Number(timeInSeconds))/3600);
  let minutes = parseInt(Number(timeInSeconds)/60)-(hours*60);
  let seconds = (Number(timeInSeconds)%60);
  let formattedSplit = timeInSeconds;


  if (timeInSeconds < 60) {
    formattedSplit = seconds.toFixed(2);
  }
  else if (timeInSeconds < 60*60) {

    if (seconds < 10) {
      formattedSplit = minutes + ":0" + seconds.toFixed(2);
    } else {
      formattedSplit = minutes + ":" + seconds.toFixed(2);
    }
  }
  else {

    formattedSplit = hours;

    if (minutes < 10) {
      formattedSplit = formattedSplit + ":0" + minutes;
    } else {
      formattedSplit = formattedSplit + ":" + minutes;
    }

    if (seconds < 10) {
      formattedSplit = formattedSplit + ":0" + seconds.toFixed(2);
    } else {
      formattedSplit = formattedSplit + ":" + seconds.toFixed(2);
    }
  }
  return formattedSplit;
}



// main display component
// calculates performance conversion using using Peter Riegel's predictor
// equation: T2 = T1 (D2/D1) * 1.06
// returns main calculation with table underneath
const ConversionResults = ({race, desiredRaceVar, hours, minutes, seconds}) => {
  
  let conversionExponent = Number(1.06);

  let currentRaceTimeVar = Number(hours)*3600 + Number(minutes)*60 + Number(seconds);

  let myNewTime = Number(currentRaceTimeVar) * (Number(desiredRaceVar)/(Number(race))) ** (conversionExponent);

  const events = [
      {
        id: 0,
        event: '400m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(400)/(Number(race))) ** (conversionExponent)),
      },
      {
        id: 1,
        event: '600m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(600)/(Number(race))) ** (conversionExponent))
      },
      {
        id: 2,
        event: '800m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(800)/(Number(race))) ** (conversionExponent)),
      },
      {
        id: 3,
        event: '1000m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(1000)/(Number(race))) ** (conversionExponent)),
      },
      {
        id: 4,
        event: '1200m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(1200)/(Number(race))) ** (conversionExponent)),
      },
      {
        id: 5,
        event: '1600m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(1600)/(Number(race))) ** (conversionExponent)),
      },
      {
        id: 6,
        event: '1609m (mile)',
        time: formatResult(Number(currentRaceTimeVar) * (Number(1609)/(Number(race))) ** (conversionExponent)),
      },
      {
        id: 7,
        event: '3000m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(3000)/(Number(race))) ** (conversionExponent)),
      },
      {
        id: 8,
        event: '3200m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(3200)/(Number(race))) ** (conversionExponent)),
      },
      {
        id: 9,
        event: '5000m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(5000)/(Number(race))) ** (conversionExponent)),
      },
      {
        id: 10,
        event: '10000m',
        time: formatResult(Number(currentRaceTimeVar) * (Number(10000)/(Number(race))) ** (conversionExponent)),
      }
    ];

return (
  <View>
    <Text style = {styles.mainResult}> { formatResult(myNewTime) } </Text>

      <Text style = {{textAlign: 'center', marginBottom: 30}}> For { desiredRaceVar }m </Text>
    <ScrollView>
      {events.map((item, index) => (
          <View key={item.id} style={styles.container}>
            <Text style={styles.text}>{item.event}</Text>
            <Text style={styles.text}>{item.time}</Text>
          </View>
        ))}
    </ScrollView>
  </View>
)



}
export default ConversionResults

const styles = StyleSheet.create ({
  container: {
      padding: 10,
      marginTop: 3,
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'flex-end',
      borderWidth: 1
   },
   text: {
      color: '#4f603c'
   },
   mainResult: {
     width: 250,
     height: 75,
     textAlign: 'center',
     padding: 18,
     fontSize: 30,
     backgroundColor: 'white',
     marginTop: 60,
     marginBottom: 10,
     marginLeft: 60
   }
})
