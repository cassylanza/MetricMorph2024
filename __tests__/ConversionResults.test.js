import React from 'react';
import { render, screen, fireEvent, getByText, toBe, queryAllByText, getByPlaceholderText } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { NavigationContainer } from '@react-navigation/native';
import { Actions } from 'react-native-router-flux'; // Mocking react-native-router-flux
import '@testing-library/jest-dom';
import ConversionResults from '../ConversionResults.js';
import { formatResult } from '../ConversionResults.js';

jest.mock('react-native-router-flux', () => ({
  Actions: {
    swimmenu1: jest.fn(),
    trackmenu1: jest.fn(),
    convperfinfoentry: jest.fn(),
    conversionresults: jest.fn(),
    splitsinfoentry: jest.fn(),
    splitsresults: jest.fn(),
    effortinfoentry: jest.fn(),
    effortresults: jest.fn(),
  },
}));

describe('ConversionResults', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<ConversionResults />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('ConversionResults component', () => {
  it('displays the main result time correctly', () => {
    const race = 1600;
    const hours = 0;
    const minutes = 6;
    const seconds = 3;
    const desiredRaceVar = 1609;

    const { getByText } = render(
      <ConversionResults race={race} hours={hours} minutes={minutes} seconds={seconds} desiredRaceVar={desiredRaceVar} />
    );

    const currentRaceTimeVar = hours*3600 + minutes*60 + seconds;
    const mySplit = currentRaceTimeVar * (desiredRaceVar / race) ** 1.06;
    const expectedMainResult = formatResult(mySplit);
    const elements = screen.queryAllByText(expectedMainResult);

    expect(elements.length).toBeGreaterThan(0);
  });
});


describe('ConversionResults component', () => {
  it('displays the distance clarifier correctly', () => {
    const { getByText } = render(
      <ConversionResults race={400} hours={0} minutes={1} seconds={3} desiredRaceVar={500} />
    );

    const distanceClarifier = getByText('For 500m');

    expect(distanceClarifier).toBeTruthy();
  });
});

  it('table appears', () => {
    const { getByText } = render(<ConversionResults />);
    expect(getByText("400m")).toBeTruthy();
  });



  describe('formatResult function', () => {
    it('returns "hh:mm:ss" for invalid input', () => {
      expect(formatResult(undefined)).toBe('hh:mm:ss');
      expect(formatResult(null)).toBe('hh:mm:ss');
      expect(formatResult('abc')).toBe('hh:mm:ss');
    });

    it('returns "hh:mm:ss" if timeInSeconds is out of range', () => {
      expect(formatResult(-1)).toBe('hh:mm:ss');
      expect(formatResult(362400)).toBe('hh:mm:ss');
    });

    it('correctly formats time less than 60 seconds', () => {
      expect(formatResult(30)).toBe('30.00');
      expect(formatResult(59.99)).toBe('59.99');
    });

    it('correctly formats time between 60 seconds and 1 hour', () => {
      expect(formatResult(65)).toBe('1:05.00');
      expect(formatResult(3600)).toBe('1:00:00.00');
    });

    it('correctly formats time greater than 1 hour', () => {
      expect(formatResult(3665)).toBe('1:01:05.00');
      expect(formatResult(3666)).toBe('1:01:06.00');
      expect(formatResult(7200)).toBe('2:00:00.00');
    });
  });


})
