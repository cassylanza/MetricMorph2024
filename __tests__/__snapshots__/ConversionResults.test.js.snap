// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionResults renders correctly 1`] = `
<View>
  <Text
    style={
      {
        "backgroundColor": "white",
        "fontSize": 30,
        "height": 75,
        "marginBottom": 10,
        "marginLeft": 60,
        "marginTop": 60,
        "padding": 18,
        "textAlign": "center",
        "width": 250,
      }
    }
  >
     
    hh:mm:ss
     
  </Text>
  <Text
    style={
      {
        "marginBottom": 30,
        "textAlign": "center",
      }
    }
  >
     For 
    m 
  </Text>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          400m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          600m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          800m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          1000m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          1200m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          1600m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          1609m (mile)
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          3000m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          3200m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          5000m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 3,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          10000m
        </Text>
        <Text
          style={
            {
              "color": "#4f603c",
            }
          }
        >
          hh:mm:ss
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
